Transition to modern JavaScript

willeastcott opened the pull request:
Inspired by [this talk](https://www.youtube.com/watch?v=cLxNdLK--yI&feature=youtu.be) from Chrome Dev Summit.

This PR migrates a portion of the PlayCanvas codebase from strict ES5 to support the very latest JS language features. Initially, the code has been updated to leverage:

* Classes
* Static class properties
* Default parameters
* `const` (sparingly for now)

Code updated so far is:

* `bundle`
* Most of `core`
* `math`
* `resources`
* `shape`
* `template`
* `extras` (MiniStats)

This is all made possible by adding Babel to the engine's rollup build process.

The rollup configuration has also been updated to generate `playcanvas.mjs`, an ES Module build of the engine. So now we have:

* Raw codebase: ES2021
* `playcanvas.mjs`: Transpiled to ES6 (ECMAScript 2015)
* `playcanvas.js`: Transpiled to ES5 (ECMAScript 2009)

This PR follows up on the [previous unmerged PR](https://github.com/playcanvas/engine/pull/2577) that migrated the engine codebase to Typescript. There was much debate about the overall benefit of adopting Typescript. Typescript provides two main benefits:

1. static type checking
2. ability to adopt the latest JS language features

But we can achieve 2 with Babel. Therefore, it is probably easier to migrate to TS-like JS (ES6 classes in particular), and then revisit the benefits of Typescript. *If* Typescript is adopted at some point, this will be a stepping stone almost directly in that direction.

Let's look at output in terms of `pc.Vec2` for example. Previously, it was:

```javascript
	function Vec2(x, y) {
		if (x && x.length === 2) {
			this.x = x[0];
			this.y = x[1];
		} else {
			this.x = x || 0;
			this.y = y || 0;
		}
	}
	Object.assign(Vec2.prototype, {
		add: function (rhs) {
			this.x += rhs.x;
			this.y += rhs.y;
			return this;
		},
```

Now it's:

```javascript
	var Vec2 = function () {
		function Vec2(x, y) {
			if (x && x.length === 2) {
				this.x = x[0];
				this.y = x[1];
			} else {
				this.x = x || 0;
				this.y = y || 0;
			}
		}

		var _proto = Vec2.prototype;

		_proto.add = function add(rhs) {
			this.x += rhs.x;
			this.y += rhs.y;
			return this;
		};
```

So essentially, a closure is introduced and `Object.assign` is removed. Still just as easy to debug. 

Feedback most definitely welcome!

Closes #2642

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

Maksims commented:
> * `playcanvas.js`: Transpiled to ES2015

I think this is mistake. As ES2015 (should be written as `ECMAScript 2015`), is ES6. And IE11 does not support ES6. It supports only ES5 (ECMA2009).

Maksims commented:
I've tested more of Babel and Closure, what it outputs, and for ES5 both seem to do a good job, but for ES6, they start to differ.

# Example 1:

Source:
```js
class Ray {
    constructor(origin = new Vec3(0, 0, 0), direction = new Vec3(0, 0, -1)) {
        this.origin = origin;
        this.direction = Object.values(direction);
    }
}
```

Closure Compiler:
```js
class Qd {
    constructor(a=new z(0,0,0), b=new z(0,0,-1)) {
        this.origin = a;
        this.direction = Object.values(b)
    }
}
```

Babel + Terser:
```js
class Ray {
    constructor() {
        var e = arguments.length > 0 && void 0 !== arguments[0] ? arguments[0] : new Vec3(0,0,0)
          , t = arguments.length > 1 && void 0 !== arguments[1] ? arguments[1] : new Vec3(0,0,-1);
        this.origin = e,
        this.direction = Object.values(t)
    }
}
```

### Issues:
1. default parameters - are converted to ES5 format
2. use of `arguments` leads to unoptimized (unable to compile to machine code) code in most JS engines.
3. it seem to not mangling classes like Closure does, which leads to higher file size

# Example 2:
Source:
```js
containsPoint(point) {
    for (const plane of this.planes) {
        if (plane[0] * point.x + plane[1] * point.y + plane[2] * point.z + plane[3] <= 0) {
            return false;
        }
    }
    return true;
}
```

Closure Compiler:
```js
containsPoint(a) {
    for (const b of this.planes)
        if (0 >= b[0] * a.x + b[1] * a.y + b[2] * a.z + b[3])
            return !1;
    return !0
}
```

Babel + Terser:
```js
containsPoint(e) {
    for (var t of this.planes)
        if (t[0] * e.x + t[1] * e.y + t[2] * e.z + t[3] <= 0)
            return !1;
    return !0
}
```

### Issues:
1. It seem to not like let/const.

# File sizes:
Right now, I've tested only on engine with small part moved to ES6+ code, but it already produces different file sizes:
* Raw ES6+ + gzip: 387Kb
* Closure Compiler ES5 + gzip: 297Kb (76.7%)
* Closure Compiler ES6 + gzip: 294Kb (76.0%)
* Babel + Terser ES5 + gzip: 287Kb (74.1%)
* Babel + Terser ES6 + gzip: 303Kb (78.3%)

Size difference between them is insignificant. ES5 Babel + Terser is smaller than Closure Compiler ES5, for around ~2.5%. While ES6 version of Babel + Terser is bigger than Closure Compiler ES6 by around ~2.3%.

Size wise, both paths produce good results, but I assume that Closure Compiler will lead for ES6, as Babael + Terser sometimes transpiles valid ES6 code to ES5.

<ins>My only concern, is why Babel + Terser is transpiling some of valid ES6 features into ES5, when instructed to use ES6 target.</ins>

willeastcott commented:
@Maksims Ha, yes, sorry, I get confused with all the JavaScript language naming. You're right - I have corrected my original post.

> Issues:
> default parameters - are converted to ES5 format

Fixed with [this commit](https://github.com/playcanvas/engine/pull/2641/commits/4fbf4a43198469673fbd3b702701385ebbd455dc).

> use of arguments leads to unoptimized (unable to compile to machine code) code in most JS engines.

I'm not seeing this. With the commit above, I'm getting the following (after beautifying):

```javascript
class Ray {
    constructor(e = new Vec3, t = new Vec3(0, 0, -1)) {
        this.origin = e, this.direction = t
    }
    set(e, t) {
        return this.origin.copy(e), this.direction.copy(t), this
    }
}
```

> it seem to not mangling classes like Closure does, which leads to higher file size

Yeah, I noticed that. I'm not sure why but I don't think it's making a significant difference.

willeastcott commented:
Oh, that the `let` + `const` thing is fixed also by that commit.

mvaligursky commented on src/core/color.js:
```diff
@@ -88,15 +88,14 @@ Object.assign(Color.prototype, {
         this.r = r;
         this.g = g;
         this.b = b;
         this.a = (a === undefined) ? 1 : a;
```
use default parameter value?

willeastcott replied:
Well spotted. Fixed.

mvaligursky commented on src/math/mat4.js:
```diff
@@ -4,6 +4,11 @@ import { Vec3 } from './vec3.js';
 import { Vec4 } from './vec4.js';
 
 var _halfSize = new Vec2();
+var x = new Vec3();
```
just curious if these could be inside class now as static or similar?

mvaligursky replied:
also, will these variables with names "x" "y" "z" create possible conflicts in the rest of the class?

willeastcott replied:
We could do. I think I'd probably put them on the `Vec3` class as `Vec3._temp1`, `Vec3._temp2`, `Vec3._temp3` maybe. But I personally don't have a big problem with these global to the module. It's kind of nice because the app code cannot access them (and therefore misuse them), which isn't true of they're on the class.

willeastcott replied:
And no, conflicts aren't possible. Rollup renames variables that conflicts when building the bundled library. It's pretty cool!

mvaligursky replied:
I mean more like conflict with a local variable inside function of Vec3. Few functions have "var x". If I don't have this var line would x reference the global x?

willeastcott replied:
Yes.

mvaligursky commented on src/math/vec2.js:
```diff
@@ -8,17 +8,45 @@
+ */
+class Vec2 {
+    constructor(x = 0, y = 0) {
+        if (x && x.length === 2) {
```
can be changed to 
`if (x.length === 2) {`

willeastcott replied:
Fixed

mvaligursky commented on src/math/vec3.js:
```diff
@@ -9,19 +9,58 @@
+ */
+class Vec3 {
+    constructor(x = 0, y = 0, z = 0) {
+        if (x && x.length === 3) {
```
can be
`if (x.length === 3) {`

willeastcott replied:
Fixed

mvaligursky commented on src/math/vec4.js:
```diff
@@ -10,21 +10,77 @@
+ */
+class Vec4 {
+    constructor(x = 0, y = 0, z = 0, w = 0) {
+        if (x && x.length === 4) {
```
same here, "x &&" part can be removed

willeastcott replied:
Fixed

mvaligursky commented on src/shape/bounding-box.js:
```diff
@@ -75,29 +74,28 @@ Object.assign(BoundingBox.prototype, {
+    copy(src) {
         this.center.copy(src.center);
         this.halfExtents.copy(src.halfExtents);
-        this.type = src.type;
```
should be ok to remove, I don't see references to it .. but curious what it was?

willeastcott replied:
So this used to be a string type property on each shape instance that specified what it was. But it was removed years ago. This bug was actually pointed out to me by the Typescript compiler but since that PR is unlikely to be merged, I'm adding the fix here.

mvaligursky commented on src/shape/oriented-box.js:
```diff
@@ -19,17 +19,16 @@ var tmpMat4 = new Mat4();
-    this._worldTransform = worldTransform.clone(); // temp - currently only used in the worldTransform accessor, see future PR for more use
-    this._aabb = new BoundingBox(new Vec3(), this.halfExtents);
-}
+        this._worldTransform = worldTransform.clone(); // temp - currently only used in the worldTransform accessor, see future PR for more use
```
should probably remove the comment, seem obsolete / irrelevant

willeastcott replied:
Removed

mvaligursky reviewed (approved):
looking great!
